metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+"""
+Sparse Compositional Metric Learning (SCML)
+"""
+
+from __future__ import print_function, absolute_import, division
+import numpy as np
+from .base_metric import _TripletsClassifierMixin, MahalanobisMixin
+from sklearn.base import TransformerMixin
+from .constraints import Constraints
+from sklearn.preprocessing import normalize
+from sklearn.neighbors import NearestNeighbors
+from sklearn.cluster import KMeans
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+from sklearn.utils import check_array, check_random_state
+import warnings
+
+# hack around lack of where in older numpy versions
+try:
+  np.sum([[0, 1], [1, 1]], where=[False, True], axis=1)
+except TypeError:
+  def sum_where(X, where):
+    return np.sum(X[where])
+else:
+  def sum_where(X, where):
+    return np.sum(X, where=where)
+
+
+class _BaseSCML_global(MahalanobisMixin):
+
+  _tuple_size = 3   # constraints are triplets
+
+  def __init__(self, beta=1e-5, basis='triplet_diffs', n_basis=None,
+               gamma=5e-3, max_iter=100000, output_iter=5000, verbose=False,
+               preprocessor=None, random_state=None):
+    self.beta = beta
+    self.basis = basis
+    self.n_basis = n_basis
+    self.gamma = gamma
+    self.max_iter = max_iter
+    self.output_iter = output_iter
+    self.verbose = verbose
+    self.preprocessor = preprocessor
+    self.random_state = random_state
+    super(_BaseSCML_global, self).__init__(preprocessor)
+
+  def _fit(self, triplets, X=None, basis=None, n_basis=None):
```

Comment:
I know, it's a bit confusing, I will try to explain it. 

This is for the case of the supervised `fit()`. In the beginning, my implementation modified `self.preprocessor` to be equal to `X`, but then I learned that `self.preprocessor`  should not be modified and that they can be different. Because of this `_prepare_inputs` doesn't have the right reference, as `triplets` are formed with reference to `X`, not `self.preprocessor`. To tackle this I choose to just pass `X` down the line so that in the future the following lines:

```
    if X is not None:
      triplets = X[triplets]
    triplets = self._prepare_inputs(triplets, type_of_inputs='tuples')
    triplets, X = self._to_index_points(triplets)
```
can be condenced in a new `_prepare_input` option like:
```
triplets, X = self._prepare_inputs(triplets, X , type_of_inputs='tuples_reference')
```
There is also another option to solve this, which is to change the reference of the triplets back to the `self.preprocessor` but for this, the reference `X` should be saved and checked before `_prepare_input` converts it to the corresponding points of the `self.preprocessor`. I found this option to be more convoluted, so I choose the other option.

I am not completely sure if this is the best way to go, so I am open for suggestions on a better way to handle this situation!

Reply:
I still think you don't need any `X`.

Why should it be different from the case of learning from pairs of quadruplets? The only thing you need in your implementation is to recover a data matrix along with triplets in the form of indices to this data matrix. While `_prepare_input` will always give you triplets in the form of features (not indices), you can easily reconstruct an data matrix and triplet indices. Even if they do not match what was given as preprocessor, who cares? This is simply for the purpose of training.

Reply:
There is no issue with that, actually `_to_index_points` doesn't recover necessarily the same `X`. The issue is that in the current way it is done if the triplets in the form of indices are passed to `_prepare_input` the reference is wrong and in consequence, the formed triplets will also be wrong!

It is perfectly possible to input the triplets already in the form of features but I am trying to think in some future new option of `_prepare_input` that will allow for a more efficient input check that doesn't require this back and forth conversion.

Reply:
Do you mean that the reason for this is to allow the user to give as input triplet indices which do not refer to the `X` given as preprocessor?

Reply:
Because otherwise I do not understand why you say "if the triplets in the form of indices are passed to `_prepare_input` the reference is wrong".

Reply:
No! In the supervised case, the `triplets` are constructed as indices references to `X` but `_prepare_input` rebuilds the triplets with the preprocessor, so if `X` and the preprocessor are not equal the reference is completely wrong.

Reply:
OK, this is for the supervised case. If I'm not mistaken, in the supervised versions of other algorithms, the tuples are always passed as a 3D array (i.e., features, not indices) to the `_fit`. This avoids the above issue, and I think this is what you should do.

Of course in the future we should find a nice way to avoid materializing the 3D array and store only indices, so that algorithms that do not need to have access to all data at once (such as this one) can benefit from lower memory complexity. But this should be done in a separate PR.

Reply:
Ok, sounds good! I will just pass the 3D array then.